### PR TITLE
add sorting and grouping channel messages selectors and test for that

### PR DESCRIPTION
**Pull request by Dinek007**

### CODE REVIEW

File: src/sagas/publicChannels/publicChannels.selectors.ts [hunk 1]
```diff
@@ -130,12 +130,102 @@ export const currentChannelDisplayableMessages = createSelector(
...
+          const dayIndex = messagesByDay.findIndex((item) => item.day === messageDay)
+          messagesByDay[dayIndex].messages.push(displayableMessage)
+        }
+      }
+    }
+    return messagesByDay
+  }
+);
+
 export const publicChannelsSelectors = {
```
Comment (siepra) [inline src/sagas/publicChannels/publicChannels.selectors.ts:229]: I wonder if you could remove selectors that are not being used in frontend from there. Also I think ZbayChannel selector can be completely deleted

File: src/sagas/publicChannels/publicChannels.selectors.ts [hunk 2]
```diff
@@ -130,12 +130,102 @@ export const currentChannelDisplayableMessages = createSelector(
...
         message: message.message,
-        createdAt: formatMessageDisplayDate(message.createdAt),
+        createdAt: message.createdAt,
         nickname: user.username,
       };
     })
 );
 
+// returns array of messages SORTED by createdAt
+export const currentChannelMessagesSortedByTime = createSelector(
```
Comment (siepra) [inline src/sagas/publicChannels/publicChannels.selectors.ts:140]: Duplicated selector. This is already being done in validCurrentChannelMessages selector

File: src/sagas/publicChannels/publicChannels.selectors.ts [hunk 3]
```diff
@@ -130,12 +130,102 @@ export const currentChannelDisplayableMessages = createSelector(
...
+  }
+);
+
+// returns array of 'day' object with day and grouped messages
+export const currentChannelMessagesGroupedByDay = createSelector(
+  currentChannelMessagesGroupedByUserTime,
+  (messages) => {
+    let messagesByDay: Array<{
+      day: string,
+      messages: any[]
```
Comment (siepra) [inline src/sagas/publicChannels/publicChannels.selectors.ts:179]: Please provide proper typing

File: src/sagas/publicChannels/publicChannels.selectors.ts [hunk 4]
```diff
@@ -130,12 +130,102 @@ export const currentChannelDisplayableMessages = createSelector(
         id: message.id,
         type: message.type,
         message: message.message,
-        createdAt: formatMessageDisplayDate(message.createdAt),
+        createdAt: message.createdAt,
```
Comment (siepra) [inline src/sagas/publicChannels/publicChannels.selectors.ts:133]: Taking under consideration the fact that next selector is a duplicate, I guess you can revert the change in this line and keep 'displayableMessage' date in readable format (which was originally the goal of this selector)

File: src/sagas/publicChannels/publicChannels.selectors.ts [hunk 5]
```diff
@@ -130,12 +130,102 @@ export const currentChannelDisplayableMessages = createSelector(
...
 
+// returns array of messages SORTED by createdAt
+export const currentChannelMessagesSortedByTime = createSelector(
+  currentChannelDisplayableMessages,
+  (messages) =>
+    messages.sort((a, b) => b.createdAt - a.createdAt)
+);
+
+// returns array of messages SORTED and GROUPED by createdAt and user
+export const currentChannelMessagesGroupedByUserTime = createSelector(
```
Comment (siepra) [inline src/sagas/publicChannels/publicChannels.selectors.ts:147]: Maybe it's just cosmetic change but I'd rather named this selector 'currentChannelMessagesMergedBySender' as it's more precise

File: src/sagas/publicChannels/publicChannels.selectors.test.ts [hunk 6]
```diff
@@ -103,46 +161,143 @@ describe('publicChannelsSelectors', () => {
...
+      },
+    ],
   },
 ]
 `);
   });
 });
 
 export {};
+// TODO wiadomosci od innych uzytkownikow
```
Comment (holmesworcester) [inline src/sagas/publicChannels/publicChannels.selectors.test.ts:303]: These should be in English so I can follow

File: src/sagas/publicChannels/publicChannels.selectors.ts [hunk 7]
```diff
@@ -130,12 +130,102 @@ export const currentChannelDisplayableMessages = createSelector(
...
+
+    for (let indexOfMessages = 0; indexOfMessages < messages.length; indexOfMessages++) {
+      let currentMessage = messages[indexOfMessages]
+
+      if (messages[indexOfMessages + 1]) {
+        while ((currentMessage.createdAt - messages[indexOfMessages + 1].createdAt) < timeOfGroupingMessages
+          && currentMessage.nickname === messages[indexOfMessages + 1].nickname) {
+          currentMessage = {
+            ...currentMessage,
+            message: currentMessage.message + "\n" + messages[indexOfMessages + 1].message
```
Comment (holmesworcester) [inline src/sagas/publicChannels/publicChannels.selectors.ts:161]: So we're still combining them into the same message? This seems like it will be a problem at some point. Is there a way we can add a flag to every message in the group (except the first) like 'grouped' that will change how it displays? If this is easy shouldn't we do it now?

**Review by holmesworcester — CHANGES_REQUESTED** [review]
Definitely change the todo text to English. Whether we change how grouped messages display is up to you. Let's do it now if it's easy.

File: src/sagas/publicChannels/publicChannels.selectors.ts [hunk 8]
```diff
@@ -130,12 +130,102 @@ export const currentChannelDisplayableMessages = createSelector(
...
+);
+
+// returns array of messages SORTED and GROUPED by createdAt and user
+export const currentChannelMessagesGroupedByUserTime = createSelector(
+  currentChannelMessagesSortedByTime,
+  (messages) => {
+    const timeOfGroupingMessages = 300
+    let newMessages = []
+
+    for (let indexOfMessages = 0; indexOfMessages < messages.length; indexOfMessages++) {
```
Comment (siepra) [inline src/sagas/publicChannels/publicChannels.selectors.ts:153]: Are you sure it'll work if there's more than 2 messages that match the condition?
 I mean, you iterate on EACH message and you check if any of the messages that comes next fits the requirements. So with three or more messages I suppose you'll bind them together with the pattern 
`{ message1: "message1 + message2 + message3" }`
then again 
`{ message2: "message2 + message3"}`

Comment (siepra) [reply src/sagas/publicChannels/publicChannels.selectors.ts:153]: Maybe it'd be more reliable if you used `reduce` method instead of `for` loop?

File: src/sagas/publicChannels/publicChannels.selectors.ts [hunk 9]
```diff
@@ -130,12 +130,102 @@ export const currentChannelDisplayableMessages = createSelector(
...
+// returns array of 'day' object with day and grouped messages
+export const currentChannelMessagesGroupedByDay = createSelector(
+  currentChannelMessagesGroupedByUserTime,
+  (messages) => {
+    let messagesByDay: Array<{
+      day: string,
+      messages: any[]
+    }> = []
+
+    for (const message of messages) {
```
Comment (siepra) [inline src/sagas/publicChannels/publicChannels.selectors.ts:182]: Similar case as above. I'm afraid of the for loop in this place, especially there is a fine designed `reduce` grouping in ZbayLite's ChannelMessages component

File: src/sagas/publicChannels/publicChannels.selectors.ts (hunk 2 above)
Comment (siepra) [reply src/sagas/publicChannels/publicChannels.selectors.ts:140]: Yet I wonder if it wouldn't be better if you first grouped the messages and then sorted those groups

File: src/sagas/publicChannels/publicChannels.selectors.test.ts (hunk 6 above)
Comment (siepra) [inline src/sagas/publicChannels/publicChannels.selectors.test.ts:303]: Please don't leave this kind of comments in the PR :D You can always to test.todo instead

File: src/sagas/publicChannels/publicChannels.selectors.ts (hunk 7 above)
Comment (siepra) [reply src/sagas/publicChannels/publicChannels.selectors.ts:161]: So the point of the selectors is to pass messages in displayable format to the frontend. The messages will remain untouched in db and also we'll be able to access them ungrouped with the help of other selectors

File: src/sagas/publicChannels/publicChannels.selectors.ts (hunk 8 above)
Comment (Dinek007) [reply src/sagas/publicChannels/publicChannels.selectors.ts:153]: for sure it works, i increasing indexOfMessages when message will group and for loop have same indexOfMessages and it will take next message
`message4` after `message1 + message2 + message3`

File: src/sagas/publicChannels/publicChannels.selectors.test.ts (hunk 6 above)
Comment (Dinek007) [reply src/sagas/publicChannels/publicChannels.selectors.test.ts:303]: yes i missed that

File: src/sagas/publicChannels/publicChannels.selectors.ts (hunk 9 above)
Comment (Dinek007) [reply src/sagas/publicChannels/publicChannels.selectors.ts:182]: as above

File: src/sagas/publicChannels/publicChannels.selectors.ts (hunk 5 above)
Comment (Dinek007) [reply src/sagas/publicChannels/publicChannels.selectors.ts:147]: agree

File: src/sagas/publicChannels/publicChannels.selectors.test.ts (hunk 6 above)
Comment (Dinek007) [reply src/sagas/publicChannels/publicChannels.selectors.test.ts:303]: yes sure, i missed it

**Review by siepra — APPROVED** [review]
I still think that this can be done better, but it seems to work so I approve

**Review by vinkabuki — APPROVED** [review]

**Review by holmesworcester — APPROVED** [review]
